### ISSUE TITLE
[Godot 4.0] - Unable to retrieve normal map from CanvasTexture atlas in TileSet/TileMap

### ISSUE THREAD

**Issue by MyStarrySpace** [godotengine/godot#61832]
### Godot version

4.0.dev (alpha 9)

### System information

Windows 11 Pro version 10.0.22000 Build 22000, NVIDIA GeForce RTX 2060, Game Ready Driver v511.79, Vulkan

### Issue description

Using a CanvasTexture with a diffuse and normal map attached will only render tiles with the diffuse map, but will not add normals to the tiles. There is seemingly no way in Godot 4.0 to use normal maps with TileSets. When tiles are pulled from the atlas to draw the tiles, normal map data should also be retrieved and added. When a user configures an atlas texture with a CanvasTexture and the TileSet is drawn, during the draw step, information about the normals can be drawn and used. Adding normals to tilesets in 3.x was possible, but there is seemingly no way in 4.0.

### Steps to reproduce

1. Create TileMap
2. Create TileSet
3. Create a tile atlas for the TileSet
4. Set the atlas texture to a new CanvasTexture
5. Add a diffuse map to the CanvasTexture
6. Add a normal map to the CanvasTexture
7. Draw some tiles onto the TileMap
8. Add a Light2D
9. Watch disappointedly as the light illuminates the tiles indiscriminately with maximum intensity

**Comment by calebmhartshorn** [comment]
I've been trying to work on this issue as it's blocking progress on my game (my first time, so slightly overwhelmed), and I noticed that shader's `NORMAL` and `NORMAL_TEXTURE` aren't working on a TileMap either, preventing easy workarounds.

Sprite2D is working as expected, and digging into the code, they both call `Texture2D::draw_rect_region()` on Ref<Texture2D>.  The only notable thing I've noticed in TileMap different from Sprite2D is that it seems to be creating lots of different CanvasItems, instead of just the one like Sprite2D.

I wonder if it's a bug with CanvasItem? Normals seem to stop working similarly in a CanvasGroup. 

Clearly don't know the code very well, so any insight would be appreciated.

**Comment by calebmhartshorn** [comment]
In the meantime, I've found a hacky workaround using a Sprite2D with a CanvasTexture and SubViewports works as expected. 

![image](https://user-images.githubusercontent.com/20672862/192147884-84ba5d18-9c1f-4773-bc77-c98830dd3408.png)

Somewhat frustrating, but it works.

**Comment by MyStarrySpace** [comment]
> In the meantime, I've found a hacky workaround using a Sprite2D with a CanvasTexture and SubViewports works as expected.
> 
> ![image](https://user-images.githubusercontent.com/20672862/192147884-84ba5d18-9c1f-4773-bc77-c98830dd3408.png)
> 
> Somewhat frustrating, but it works.

How did you manage to get the contents of the SubViewport to display in the sprite? Did you use a ViewPort texture as the sprite texture? Do the viewports need any special settings?

This is a much better workaround than what I've currently been doing, which is to copy the normals and diffuse onto a sprite, then have it on the Premultiplied Alpha blending mode on top of the tileset.

**Comment by calebmhartshorn** [comment]
Yes, the local to scene property is in the resources group of the CanvasTexture.
![image](https://user-images.githubusercontent.com/20672862/193263584-814ae627-62a3-41e1-b476-456a19f6225e.png)

Couple caveats with this whole approach: 
- You need two copies of your tileset. One with diffuse textures and one with normal textures (plus specular if you want it).
- With every change you make to one TileMap, you need to copy paste to the other to avoid lighting artifacts.
- There's a few bugs with SubViewports at the moment. If you get glitchy or missing textures, try changing the SubViewport size to something other than the defualt, set their Render Target Update Mode to always, and/or just reopen the project.

Here's an example project I made for you this morning, in case you get stuck:

[NormalTilemaps.zip](https://github.com/godotengine/godot/files/9683791/NormalTilemaps.zip)

Due to the issues mentioned above, I would **strongly** recommend writing a script to automatically replace your TileMap with this setup at runtime.

**Comment by mrcdk** [comment]
Looks like the issue comes from enabling `Use texture padding` under the atlas properties of the tileset which creates a new `ImageTexture` from whatever source texture has been assigned. (https://github.com/godotengine/godot/blob/master/scene/resources/tile_set.cpp#L4587) Disabling that property fixes the issue.

![Godot_v4 0-beta3_win64_8GYsOHx8jn](https://user-images.githubusercontent.com/483531/198866488-12402ded-8e61-4c4c-8673-f0cdbb28a704.gif)

**Comment by groud** [comment]
> Looks like the issue comes from enabling Use texture padding under the atlas properties of the tileset which creates a new ImageTexture from whatever source texture has been assigned. ([master/scene/resources/tile_set.cpp#L4587](https://github.com/godotengine/godot/blob/master/scene/resources/tile_set.cpp?rgh-link-date=2022-10-30T06%3A59%3A55Z#L4587)) Disabling that property fixes the issue.

Oh right, that makes a lot of sense! Support for CanvasTexture then has to be added to the texture padding feature.
